Vizzuality/marxan-cloud
# API - authentication

### Overview

This PR adds authentication support to the API.

Users can log in via username (email) and password.

`UsersController` and `ProjectsControllers` are now protected via `JwtAuthGuard`

We also have:

* issuance of a fresh JWT upon presenting a still-valid one for the same user (`POST /api/v1/auth/refresh-token`)
* logging of issued JWTs
* checking that JWT being presented for authentication are not revoked 🚀 
* an endpoint (`GET /api/v1/users/me`) to retrieve metadata of the current user

I have also snuck in an unrelated change (sorry 🙄) - [`typescript-coverage-report`](https://github.com/alexcanessa/typescript-coverage-report) as I didn't want to create a tiny PR just for this (laame excuse!).

### Designs

N/A

### Testing instructions

Reset API db and create test users (the test seed has been updated in this PR)

```
make clean-slate
make start
make seed-api-with-test-data
```

Try to authenticate as one of the test users:

```
curl -X POST -H 'Content-Type: application/json' http://localhost:3030/auth/login -d '{"username": "aa@example.com", "password": "aauserpassword"}'
```
A JWT token should be returned by the API.

Try to create a new user:

```
curl -X POST -H 'Content-Type: application/json' http://localhost:3030/auth/sign-up -d '{"username": "dd@example.com", "password": "dduserpassword"}'
```

The new user should be listed alongside the other test ones:

```
curl http://localhost:3030/api/v1/users
```

Try authenticating as the new user:

```
curl -X POST -H 'Content-Type: application/json' http://localhost:3030/auth/login -d '{"username": "dd@example.com", "password": "dduserpassword"}'
```

A JWT token should be returned

### Feature relevant tickets

https://www.pivotaltracker.com/story/show/176376651

---

## Checklist before submitting

      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
